enhancements/installer/operators-check.md

```diff
@@ -0,0 +1,174 @@
+an operator never stops progressing so that I can triage operator bugs.
+
+### Goals
+
+* Installer correctly identifies a failed cluster install where cluster operators are not stable
+* Installer exits non-zero in those situations and delivers meaningful error message
+
+### Non-Goals
+
+* Installer handling other classes of errors, such as failure to provision machines by MAO
+
+## Proposal
+
+### Workflow Description
+
+Cluster admin will begin installing cluster as usual. The Installation workflow will be:
+
+1. Cluster initializes as usual
+2. As usual, installer checks that cluster version is Available=true Progressing=False Degraded=False 
+3. Installer checks status of each cluster operator for stability
+4. If a cluster operator is not stable the installer logs a message and throws an error
+
+### API Extensions
+
+None
+
+### Implementation Details/Notes/Constraints [optional]
+
+The important detail is how we define _stable_. The definition proposed here is:
+
+```
+If a cluster operator does not maintain Progressing=False for at least 30 seconds,
+during a five minute period it is unstable.
+```
+
+### Risks and Mitigations
+
+One risk would be a false positive: the Installer identifies that a cluster
+operator is unstable, but it turns out the operator is perfectly healthy;
+the install was declared a failure but was actually successful. This risk
+seems low and a risk that could be managed by monitoring these specific failures
+in CI.
+
+### Drawbacks
+
+This enhancement has some significant potential drawbacks in terms of design:
+
+A fundamental design principle of 4.x and the operator framework has been
+delegating responsibility away from the Installer to allow clean separation
+of concerns and better maintainbility. Without this enhancement, it is the
+responsibility of the Cluster-Version Operator to determine whether given
+Cluster Operators constitute a successful version. The idea of keeping the
+cluster-version operator as the single responsible party is discussed in the
+alternatives section.
+
+Following the previous point, this enhancement could introduce a slippery slope.
+Where before this enhancement, the Installer would simply query the CVO; now we
+are adding additional conditional logic. Introducing this conditional logic leads
+to the risk of further conditions, which could make the install process more brittle
+and introduce the potential for false positives or other failures.
+
+Does implementing this enhancement address symptoms of issues with operator status definitions?
+Should operators themselves be setting Degraded=True when they don't meet this stability criterion?
+
+As we have seen with other timeouts in the Installer, developers and users will want to change these.
+We should define a process for refining our stability criteria. 
+
+## Design Details
+
+### Open Questions [optional]
+
+1. What is the correct definition of a stable operator? More importantly, how can we
+refine this definition?
```

Comment:
I think it's progressing=False for the last 30 seconds but only waiting for all COs to achieve that up to 5 minutes. Only waiting up to five minutes because these aren't failed installs, the cluster is in all cases I'm aware of, a viable cluster that just differs from the exact spec defined in install-config.

Reply:
> Is the installer going to look at any other conditions on the operator, e.g. available=false, or it will continue to rely on the CVO to proxy that?

No, it would continue to rely on CVO as a proxy which is a good way of putting it. 

Installer checks for bootstrap complete configmap. If that's good:
Installer checks CV. If that's good:
Installer sets a 5 minute context deadline then watches each CO to ensure they achieve Progress=False with LastTransition time > 30 seconds If that's good,
Installer checks console availability

I think that's it.

Reply:
I'm happy with the definition as put forward by this enhancement plus the existing logic that requires A=t and D=f.